File: src/cpp/src/EXAMPLE/minijob.cc

#include "lcio.h"

#include "IO/LCWriter.h"
#include "EVENT/LCIO.h"
#include "DATA/LCFloatVec.h"

#include "IMPL/LCEventImpl.h" 
#include "IMPL/LCRunHeaderImpl.h" 
#include "IMPL/LCCollectionVec.h"
#include "IMPL/SimCalorimeterHitImpl.h"
//#include "IMPL/SimTrackerHitImpl.h"
#include "IMPL/MCParticleImpl.h" 
#include "IMPL/LCFlagImpl.h" 
#include "IMPL/LCTOOLS.h"

#include <cstdlib>
#include <iostream>


using namespace std ;
using namespace lcio ;

static const int NRUN = 1 ;
static const int NEVENT = 1 ; // events
static const int NMCPART = 1 ;  // mc particles per event
static const int NHITS = 1 ;  // calorimeter hits per event

static string FILEN = "minijob.sio" ;


/** Simple test program to demonstrate writing of data with lcio.
 * mininmal data file for debugging
 */

int main(int argc, char** argv ){
  
  // read file name from command line (only argument) 
  if( argc > 1 ) {
    FILEN = argv[1] ;
  }
  
  // create sio writer
  LCWriter* lcWrt = LCFactory::getInstance()->createLCWriter() ;
  
  try{  lcWrt->open( FILEN ) ;
  }
  catch( IOException& e ){    cout << e.what() << endl ;
    return 0 ;
  }

  // loop over runs
  for(int rn=0;rn<NRUN;rn++){
    
    LCRunHeaderImpl* runHdr = new LCRunHeaderImpl ; 
    runHdr->setRunNumber( rn ) ;
    
    string detName("MiniDet")  ;
    runHdr->setDetectorName( detName ) ;

    string description("minimal file") ;
    runHdr->setDescription( description ) ;
    
    string ecalName("ECALmini") ;
    runHdr->addActiveSubdetector( ecalName ) ;

    lcWrt->writeRunHeader( runHdr ) ;

    // EventLoop - create some events and write them to the file
    for(int i=0;i<NEVENT;i++){
      
      // we need to use the implementation classes here 
      LCEventImpl*  evt = new LCEventImpl() ;
      
      
      evt->setRunNumber(  rn   ) ;
      evt->setEventNumber( i ) ;
      evt->setDetectorName( detName ) ;
      
      // create and add some mc particles 
      LCCollectionVec* mcVec = new LCCollectionVec( LCIO::MCPARTICLE )  ;
      
      MCParticle* mom = 0 ;
      for(int j=0;j<NMCPART;j++){
	MCParticleImpl* mcp = new MCParticleImpl ;
	
	mcp->setPDG( 101 + j*100 ) ;
	mcp->setParent(  mom  );
	
	float p[3] = { 2.*1024. , 4.*1024. , 8.*1024. } ;
	mcp->setMomentum( p ) ;
	
	mom =  mcp ;  // one body decays :-)
	mcVec->push_back( mcp ) ;
      }
      
      // now add some calorimeter hits
      LCCollectionVec* calVec = new LCCollectionVec( LCIO::SIMCALORIMETERHIT )  ;
      
      // set flag for long format (including position )
      // and PDG 
      LCFlagImpl chFlag(0) ;
      chFlag.setBit( LCIO::CHBIT_LONG ) ;
      chFlag.setBit( LCIO::CHBIT_PDG ) ;
      calVec->setFlag( chFlag.getFlag()  ) ;


      for(int j=0;j<NHITS;j++){
	
	SimCalorimeterHitImpl* hit = new SimCalorimeterHitImpl ;
	
	hit->setEnergy( 3.1415 )  ;
	
	float pos[3] = { 1. , 2. , 4. } ;
	
	hit->setPosition( pos ) ;
	
	calVec->push_back( hit ) ;
	

	hit->addMCParticleContribution(  dynamic_cast<MCParticle*>(mcVec->getElementAt(0)) , 
					 1. , 2. , 4 ) ; 
	
      }
      
      evt->addCollection( (LCCollection*) mcVec , "MCParticle" ) ;
      evt->addCollection( (LCCollection*) calVec , ecalName ) ;
      
      // dump the event to the screen 
      LCTOOLS::dumpEvent( evt ) ;
      
      // write the event to the file
      lcWrt->writeEvent( evt ) ;
      
      // we created the event so we need to take care of deleting it ...
      delete evt ;
      
    } // evt loop
  } // run loop
    
  cout << endl 
       << " created  " << NRUN << " runs with  " << NRUN*NEVENT << " events"  
       << endl << endl ;

  lcWrt->close() ;
  
  return 0 ;
}

File: src/java/hep/lcio/example/MiniJob.java
package hep.lcio.example;

import hep.lcio.event.*;

import hep.lcio.implementation.event.*;
import hep.lcio.implementation.io.LCFactory;

import hep.lcio.io.*;

import java.io.IOException;
import java.util.Random;


/**
 *
 * @author Tony Johnson
 * @version $Id: MiniJob.java,v 1.5 2003-11-08 03:08:50 tonyj Exp $
 */
public class MiniJob
{
   private final static int NRUN = 1;
   private final static int NEVENT = 1;
   private final static int NMCPART = 1; // mc particles per event
   private final static int NHITS = 1; // calorimeter hits per event
   private final static String detName = "MiniDet";
   private final static String ecalName = "ECALmini";

   /**
    * @param args the command line arguments
    */
   public static void main(String[] args)
   {
      // create sio writer
      Random random = new Random();
      LCWriter lcWrt = LCFactory.getInstance().createLCWriter();

      try
      {
         lcWrt.open(args[0]);
      }
      catch (IOException e)
      {
         System.out.println("cannot open file" + args[0] + e.getMessage());
         System.exit(1);
      }

      // loop over runs
      for (int rn = 0; rn < NRUN; rn++)
      {
         ILCRunHeader runHdr = new ILCRunHeader();
         runHdr.setRunNumber(rn);
         runHdr.setDetectorName(detName);
         runHdr.setDescription("minimal file");
         runHdr.addActiveSubdetector(ecalName);

         try
         {
            lcWrt.writeRunHeader(runHdr);
         }
         catch (IOException e)
         {
            System.out.println("Couldn't write run header " + rn);
         }

         // EventLoop - create some events and write them to the file
         for (int i = 0; i < NEVENT; i++)
         {
            // we need to use the implementation classes here
            ILCEvent evt = new ILCEvent();

            evt.setRunNumber(rn);
            evt.setEventNumber(i);
            evt.setDetectorName(detName);

            // create and add some mc particles
            ILCCollection mcVec = new ILCCollection(LCIO.MCPARTICLE);

            IMCParticle mom = null;
            for (int j = 0; j < NMCPART; j++)
            {
               IMCParticle mcp = new IMCParticle();

               mcp.setPDG(101 + (j * 100));
               mom.addDaughter(mcp);

               float[] p = { 2.f * 1024.f, 4.f * 1024.f, 8.f * 1024.f };
               mcp.setMomentum(p);

               mom = mcp; // one body decays :-)
               mcVec.add(mcp);
            }

            // now add some calorimeter hits
            ILCCollection calVec = new ILCCollection(LCIO.SIMCALORIMETERHIT);

            int flag = 1 << LCIO.CHBIT_LONG; //
            flag = flag | (1 << LCIO.CHBIT_PDG);
            calVec.setFlag(flag);

            for (int j = 0; j < NHITS; j++)
            {
               ISimCalorimeterHit hit = new ISimCalorimeterHit();
               hit.setEnergy(3.1415f);

               float[] pos = { 1.f, 2.f, 4.f };

               hit.setPosition(pos);

               calVec.add(hit);

               hit.addMCParticleContribution((MCParticle) mcVec.getElementAt(0), 1.f, 2.f, 4);
            }

            // add all collection to the event
            evt.addCollection(mcVec, "MCParticle");
            evt.addCollection(calVec, ecalName);

            // dump the event to the screen
            LCTools.dumpEvent(evt);

            // write the event to the file
            try
            {
               lcWrt.writeEvent(evt);
            }
            catch (IOException e)
            {
               System.out.println("Couldn't write event " + i);
            }
         }

         // evt loop
      }

      // run loop
      System.out.println();
      System.out.println(" created  " + NRUN + " runs with  " + (NRUN * NEVENT) + " events");
      try
      {
         lcWrt.close();
      }
      catch (Exception e) {}
   }
}
